src/GeniusYield/Types/TxMetadata.hs

```diff
@@ -5,110 +5,125 @@ License     : Apache 2.0
--- Convenience functions for dealing with 'GYTxMetadataValue's and 'TxMetadata'
---------------------------------------------------------------------------------
-
--- | Apply 'Map' constructor of 'GYTxMetadataValue'.
-gyTxMetaMap :: [(GYTxMetadataValue, GYTxMetadataValue)] -> GYTxMetadataValue
-gyTxMetaMap = Api.TxMetaMap
-
--- | Apply 'List' constructor of 'GYTxMetadataValue'.
-gyTxMetaList :: [GYTxMetadataValue] -> GYTxMetadataValue
-gyTxMetaList = Api.TxMetaList
-
--- | Apply 'Number' constructor of 'GYTxMetadataValue'.
-gyTxMetaNumber :: Integer -> GYTxMetadataValue
-gyTxMetaNumber = Api.TxMetaNumber
-
--- | Apply 'Bytes' constructor of 'GYTxMetadataValue'.
-gyTxMetaBytes :: ByteString -> GYTxMetadataValue
-gyTxMetaBytes = Api.TxMetaBytes
-
--- | Apply 'Text' constructor of 'GYTxMetadataValue'.
-gyTxMetaText :: Text -> GYTxMetadataValue
-gyTxMetaText = Api.TxMetaText
-
--- | Create a 'GYTxMetadataValue' from a Text as a list of chunks of an acceptable size.
-gyMetaTextChunks :: Text -> GYTxMetadataValue
-gyMetaTextChunks = Api.metaTextChunks
-
--- | Create a 'GYTxMetadataValue' from a 'ByteString' as a list of chunks of an accaptable size.
-gyMetaBytesChunks :: ByteString -> GYTxMetadataValue
-gyMetaBytesChunks = Api.metaBytesChunks
+import qualified Cardano.Api                           as Api
+import           Data.ByteString                       (ByteString)
+import qualified Data.ByteString                       as BS
+import qualified Data.Map.Strict                       as Map
+import qualified Data.Text.Encoding                    as TE
+import           Data.Word                             (Word64)
+import           GeniusYield.Imports                   (Text, foldl')
+import           GeniusYield.Types.TxMetadata.Internal (GYTxMetadataValue (..),
+                                                        txMetadataValueFromApi,
+                                                        txMetadataValueToApi)
+
+
+newtype GYTxMetadata = GYTxMetadata (Map.Map Word64 GYTxMetadataValue)
+  deriving newtype (Eq, Semigroup, Show)
+
+txMetadataToApi :: GYTxMetadata -> Api.TxMetadata
+txMetadataToApi (GYTxMetadata m) = Api.TxMetadata (Map.map txMetadataValueToApi m)
+
+txMetadataFromApi :: Api.TxMetadata -> GYTxMetadata
+txMetadataFromApi (Api.TxMetadata m) = GYTxMetadata (Map.map txMetadataValueFromApi m)
+
+-- | Construct a 'GYTxMetadataValue' from a list of key-value pairs.
+constructTxMetadataMap :: [(GYTxMetadataValue, GYTxMetadataValue)] -> GYTxMetadataValue
+constructTxMetadataMap = GYTxMetaMap
+
+-- | Construct a 'GYTxMetadataValue' from a list of 'GYTxMetadataValue's.
+constructTxMetadataList :: [GYTxMetadataValue] -> GYTxMetadataValue
+constructTxMetadataList = GYTxMetaList
+
+-- | Construct a 'GYTxMetadataValue' from an 'Integer'. Returning 'Nothing' if the given 'Integer' is not in the range @-(2^64-1) .. 2^64-1@.
+constructTxMetadataNumber :: Integer -> Maybe GYTxMetadataValue
```

Comment:
Idea is to provide safer API, as in this function so that `GYTxMetadata` is correct by construction. For internal usage, `GYTxMetadata.Internal` module is provided.